### PR TITLE
Improved "find_substr" implementations

### PR DESCRIPTION
**Pull request by ykalchevskiy**
Fixed index error
Provided more "pythonic" implementation

### CODE REVIEW

**Comment by ykalchevskiy** [comment]
Hi, @1st!
Any chances that you will merge this pull request?

**Comment by ykalchevskiy** [comment]
@1st, m? :)

**Comment by 1st** [comment]
Hey. I'm sorry, didn't have time to review it earlier. Will do it today evening. Thanks for notifications!

File: algorithms.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ code here
 ```python
 # general language-independent implementation
 def find_substr(string, substr):
-    for i in range(0, len(string)):
+    for i in range(0, len(string) - len(substring) + 1):
```
Comment (1st) [inline algorithms.md]: It looks like if we have two strings of lenght zero - why do we need to add this extra one mumber to have an empty iteration over empty strings?

File: algorithms.md [hunk 2]
```diff
@@ -38,10 +38,7 @@ def find_substr(string, substr):
 ```python
 # pythonic implementation with string slice
 def find_substr(string, substr):
-    for i in range(0, len(string)):
-        if substr == string[i:i + len(substr)]:
-            return True
-    return False
+    return any(substr == string[i:i + len(substr)] for i in range(len(string) - len(substr) + 1))
```
Comment (1st) [inline algorithms.md]: The same here

**Review by 1st — CHANGES_REQUESTED** [review]

File: algorithms.md (hunk 1 above)
Comment (ykalchevskiy) [reply algorithms.md]: Of course we can add an extra step like `if substr == '': return True`, but the current code also checks this. Without an iteration we can not reach `return True`.
This `+1` is checked by a case with two non-empty strings with same length: we really need one iteration to compare them.

BTW, there is a typo in the patch: I used `substring` but `substr` is right. I will update it soon.

**Comment by 1st** [comment]
Hi @ykalchevskiy.

Did you test your solution on edge cases (empty string, no substring found, substr is equal to original srting)?

**Comment by ykalchevskiy** [comment]
Hi @1st,
Sure. We can add tests somewhere.

``` python
import pytest


def find_substr(string, substr):
    for i in range(0, len(string) - len(substr) + 1):
        for j in range(0, len(substr)):
            if substr[j] != string[i + j]:
                break
        else:
            return True
    return False


def find_substr_py(string, substr):
    return any(substr == string[i:i + len(substr)] for i in range(len(string) - len(substr) + 1))


T = True
F = False


@pytest.mark.parametrize('string substr result'.split(), [
('', '', T),
('', 'a', F),
('a', '', T),
('a', 'a', T),
('a', 'aa', F),
('aa', 'a', T),
('abc', 'abc', T),
('abcd', 'abc', T),
('ababc', 'abc', T),
('ababcd', 'abc', T),
('abcabc', 'abcd', F),
])
@pytest.mark.parametrize('func', [find_substr, find_substr_py])
def test_(string, substr, result, func):
    assert func(string, substr) is result
```

And output:

``` bash
$ py.test -q test_.py
......................
22 passed in 0.02 seconds
```

**Comment by 1st** [comment]
Good. Can you please create teo folders - first for samples, and second for tests. And place code and tests in appropriate folders - "code" and "tests". Thank you

**Comment by ykalchevskiy** [comment]
Hi @1st,

I think it would be better just to add tests as doctests as in the latter example 'permute_string' and in [the Python examples](https://github.com/1st/interview/blob/master/python.md#write-test-for-function). Do you agree?

**Comment by 1st** [comment]
Yes, please do it in docstring. Good idea.

File: algorithms.md (hunk 2 above)
Comment (ykalchevskiy) [reply algorithms.md]: Answered.

**Comment by ykalchevskiy** [comment]
Done :)

**Comment by ykalchevskiy** [comment]
Hi @1st,

What about merge? :)

**Comment by 1st** [comment]
Still waiting for response on comments in code. At least I see unresolved two comments - add comment to a function and don't use `any` by described reason.

**Comment by ykalchevskiy** [comment]
Oh, unfortunately I don't see any unanswered comments :( Where are they?

**Comment by 1st** [comment]
Scroll above on 4 comments to see two unresolved coments. Comments to file in this trhread of discussion.

**Comment by ykalchevskiy** [comment]
I see only these comments, but I have already answered.

**Comment by 1st** [comment]
Hi Yan.

I see this comments unresolved:

screen shot 2016-12-01 at 11 39 37 am

Also note that I have moved all chapters to the chapters/ directory. Sync before you will make final changes. Thank you.

**Comment by ykalchevskiy** [comment]
Hi, @1st,

Unfortunately your review is **pending**: it means these comments are **only visible to you**!

https://help.github.com/articles/reviewing-proposed-changes-in-a-pull-request/#reviewing-the-proposed-changes-in-line-comments

**Comment by ykalchevskiy** [comment]
Anyway, about ```any```: actually it will iterate up to first element for which ```bool(el) is True```. It states in the docs: https://docs.python.org/2/library/functions.html?highlight=any#any
Also we can check it using something like this:
```python
def f(x):
  print x
  return x
any(f(i) for i in [1, 0]) 
```
Only ``1`` will be printed.

**Comment by ykalchevskiy** [comment]
I just added a description and rebased changes to fit updated structure.

**Comment by ykalchevskiy** [comment]
Hi @1st,

Let's finish it. What do you think?

**Comment by ykalchevskiy** [comment]
Hi @1st,

Any comments? :)

**Comment by ykalchevskiy** [comment]
Hi @1st,

One more time...

**Comment by 1st** [comment]
Sorry for the long ignoring. It's a lot of work here. Congratulate you with the first push to this repository!
